lib/fsm/utils.go

```diff
@@ -43,9 +43,85 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 		return trace.BadParameter(
 			"phase %q has already been rolled back", phase.ID)
 	}
+
+	// TODO: Rollback of non-leaf phases is not currently supported.
+	// Rollback starts top-down, and not in reverse order.
+	if phase.HasSubphases() {
+		return trace.BadParameter(
+			"unable to rollback non-leaf phase, only leaf phases can be rolled back").AddField("phase", phase.ID)
+	}
+
+	if !dependentsRolledBack(plan, phase.ID) {
+		return trace.BadParameter(
+			"rollback subsequent phases before rolling back phase %q", phase.ID)
+	}
 	return nil
 }
 
+// dependentsRolledBack returns true if all phases that are dependent on the
+// phase specified by phaseID is unstarted or has been rolled back.
+func dependentsRolledBack(plan *storage.OperationPlan, phaseID string) bool {
+	// required will be nil if an invalid phaseID is provided.
+	required := getRequired(plan.Phases, phaseID)
+	if required == nil {
+		return true
+	}
+	return dependentsRolledBackHelper(required, plan.Phases)
+}
+
+// getRequired constructs the initial set of required phases. This set includes
+// the phase specified by phaseID and its parent phases.
+func getRequired(phases []storage.OperationPhase, phaseID string) map[string]struct{} {
```

Comment:
I would update the godoc comment with an example of what it returns, to simplify understanding. I.e.

```
Given plan like:

/init
/masters
  * /node-1
    * /system-upgrade
  * /node-2

for "/masters/node-1/system-upgrade" phase id it will return:

["/masters", "/masters/node-1", "/masters/node-1/system-upgrade"]
```

Also, why does this return a map, doesn't the order matter?

Reply:
Order doesn't matter, we just need the set to check if any other phases 'requires' any of the phases in this set.